### ISSUE TITLE
bundle install is stuck (bundler v2.2.15)

### ISSUE THREAD

**Issue by madejejej** [rubygems/rubygems#4539]
### Describe the problem as clearly as you can

This seems similar to https://github.com/rubygems/rubygems/issues/4240 but that issue is marked as fixed.

`bundle install` cannot complete with the Gemfile below and is instead stuck on `Resolving dependencies.....` step.

### Post steps to reproduce the problem

- Ruby 2.6.5
- Bundler 2.2.15
- Gem 3.0.3

<details>
<summary>
Gemfile
</summary>

```ruby
# frozen_string_literal: true

source 'https://rubygems.org'

git_source(:github) { |repo| "https://github.com/#{repo}.git" }

gem 'activesupport'
gem 'actionpack'
gem 'actionview'
gem 'activemodel'
gem 'activerecord'
gem 'actionmailer'
gem 'railties'
gem 'sprockets-rails'

gem 'ace-rails-ap'
gem 'activeadmin'
gem 'activeadmin_json_editor'
gem 'activerecord-import'
gem 'analytics-ruby', '2.2.5'
gem 'attr_encrypted'
gem 'aws-sdk-applicationautoscaling'
gem 'aws-sdk-dynamodb'
gem 'aws-sdk-s3'
gem 'avromatic'
gem 'bootsnap', require: false
gem 'chargebee'
gem 'chartkick'
gem 'config'
gem 'connection_pool'
gem 'ddtrace'
gem 'dogstatsd-ruby', require: 'datadog/statsd'
gem 'dotenv-rails', '2.7.1'
gem 'dry-configurable', '0.11.5'
gem 'dry-container'
gem 'dry-events'
gem 'dry-initializer'
gem 'dry-struct'
gem 'dry-transaction'
gem 'dry-types'
gem 'dry-validation'
gem 'elasticsearch'
gem 'elasticsearch-model'
gem 'faraday'
gem 'fast_blank'
gem 'flipper'
gem 'flipper-active_record'
gem 'flipper-ui'
gem 'foundation_emails'
gem 'grape', '1.4.0'
gem 'grape-entity'
gem 'haml'
gem 'haml-rails'
gem 'hashie'
gem 'hiredis'
gem 'hive_geoip2'
gem 'ios-devices'
gem 'iso_country_codes'
gem 'lru_redux'
gem 'murmurhash3'
gem 'mustache'
gem 'oj'
gem 'ped'
gem 'pg'
gem 'pghero'
gem 'phonelib'
gem 'premailer-rails'
gem 'rack'
gem 'rack-utf8_sanitizer'
gem 'redis', require: %w[redis redis/connection/hiredis]
gem 'redis-namespace'
gem 'redlock'
gem 'request_store'
gem 'request_store-sidekiq'
gem 'rollbar'
gem 'semian', require: %w[semian/redis]
gem 'sequel-activerecord_connection'
gem 'slack-notifier'
gem 'sqlite3'
gem 'stripe'
gem 'typhoeus'
gem 'user_agent_parser'

gem 'rdkafka'
gem 'waterdrop'

gem 'whenever', require: nil

# former assets
gem 'bootstrap-sass'
gem 'mini_racer'
gem 'sass-rails'
gem 'uglifier', require: nil

group :development do
  gem 'listen'
  gem 'web-console'
  gem 'spring'
  gem 'spring-commands-rspec'
  gem 'spring-watcher-listen'
end

group :development, :test do
  gem 'active_record_query_trace'
  gem 'awesome_print'
  gem 'byebug'
  gem 'rspec-rails'
  gem 'toxiproxy' 
end

group :test do
  gem 'cucumber-rails', require: false
  gem 'database_cleaner'
  gem 'factory_bot_rails'
  gem 'faker'
  gem 'json_expressions'
  gem 'rspec-collection_matchers'
  gem 'simplecov'
  gem 'timecop'
  gem 'webmock', require: nil
end

group :production do
  gem 'puma'
end
```

</details>

### Which command did you run?

`DEBUG_RESOLVER=1 bundle install --verbose `

### What were you expecting to happen?

The command should finish.

### What actually happened?

The command did not finish even after an hour.

**Comment by duckinator** [comment]
I'm currently running it in the background, and I'll check where it is after an hour and again after 2-4 hours. If it doesn't finish in that time, the next step is to try to determine if there's a loop.

If there is a loop, the next step is to try to find a simplified testcase. If there's not a loop, I think the next step is determining whether there is a specific situation causing performance problems, or if it's just so big of a list of possibilities that it takes a long time.

**Comment by duckinator** [comment]
> If it doesn't finish in that time, the next step is to try to determine if there's a loop.

It ran for over 2 days because I got caught up with other things and forgot to stop it.

@dwelch-r7 thank you for the information. I'm trying @madejejej's Gemfile with 2.1.4 to compare what happens. metasploit_data_models appears to be a lot simpler, and might work as a good basis for a testcase. The fact that yours finishes also confirms that yours doesn't result in a loop, so that's good to know as well.

**Comment by duckinator** [comment]
Assuming these are the same issue, I think I may have a testcase based on metasploit_data_models. I'm working on simplifying it, now.

**Comment by duckinator** [comment]
Testcase as it is now takes 40 seconds on bundler 2.1.4 (including building some C extensions), but with 2.2.17 it took over 10 minutes and still hadn't finished.

**Comment by duckinator** [comment]
Hey &mdash; sorry for the delay here. Had a lot of stuff come up at once. Planning to get back to this in the next day or two. :slightly_smiling_face: 

I forgot to post the testcase I mentioned above, so I'll include it below. I still have some work to do on narrowing down the exact cause of the problem, but it's significantly smaller than the original examples.

```ruby
source 'https://rubygems.org'

gem 'rails', '4.2.5.1'
gem 'bootstrap-sass', '~> 3.3', '>= 3.3.6'
gem 'sass-rails', '~> 5.0'
gem 'uglifier', '>= 1.3.0'
gem 'coffee-rails', '~> 4.1.0'
gem 'jquery-rails'
gem 'turbolinks'
gem 'jbuilder', '~> 2.0'
gem 'sdoc', '~> 0.4.0', group: :doc

group :development, :test do
  gem 'byebug'
end

group :development do
  gem 'web-console', '~> 2.0'
  gem 'sqlite3'
  gem 'spring'
end

group :production do
#  gem 'pg',             '0.17.1'
#  gem 'rails_12factor', '0.0.2'
 # gem 'puma',           '3.1.0'
end
```

**Comment by duckinator** [comment]
(or 3, possibly, because nothing throws off productivity like it hitting 83°F _inside_ your house and seeing if you can get air conditioning installed sooner. :sweat_smile: )

**Comment by duckinator** [comment]
I am extremely confused because I came back to this and now running `DEBUG_RESOLVER=1 bundle install --verbose` tells me this:

```
Bundler::VersionConflict: Bundler could not find compatible versions for gem "bundler":                                
  In Gemfile:
    rails (= 4.2.5.1) was resolved to 4.2.5.1, which depends on                                                        
      bundler (>= 1.3.0, < 2.0)                                                                                        
                                                                                                                       
  Current Bundler version:                                                                                             
    bundler (2.2.17)      
```

And now I have NO IDEA how this testcase ever worked.

**Comment by duckinator** [comment]
I'm a dingus and accidentally pasted the wrong file before. I found the correct one and simplified it a bit more. :sweat_smile: 

Here's the actual testcase: https://github.com/duckinator/rubygems-4539-testcase

**Comment by duckinator** [comment]
Poking at this a bit more today.

With the current testcase:

- Bundler 2.1.4 finishes the resolution with 172 steps (0.3 seconds).
- Bundler 2.2.17 does not finish the resolution after 17,720 steps (324.6 seconds).

It has given itself _two orders of magnitude_ more work to do and takes _three orders of magnitude_ longer to do it.

I'm _by no means an expert_ in this area so I may be wrong, but my understanding of the problem is:

1. Dependency resolution is a [topological sorting](https://en.wikipedia.org/wiki/Topological_sorting) problem.
2. Topological sorting typically has running time that scales linearly with the _number of nodes plus the number of edges._
3. In this case, the nodes are the gems and the edges are the version constraints.
4. The information I found above implies we have added significantly more edges, causing a kind of explosion in complexity.

_If_ this is correct, there's 3 main solutions to the problem:

1. Reduce the number of nodes (a project's dependencies).
2. Reduce the number of edges (internal Bundler stuff).
3. Use a faster algorithm.

However, the fact that my testcase reproduces this problem with only _7 direct dependencies_ means that the first option is not a good solution in some cases.

**Comment by deivid-rodriguez** [comment]
The resolver on 2.2 is more complex since it also considers platforms for resolution. That said, it seems much more correct now, not only because of platform resolution bugs having been fixed, but also because issues like this one seem less frequent now than on bundler 2.1. As a matter of fact, I believe this is the only "infinite resolution" issue left open whereas we used to have plenty of them before which are now fixed.

The usual workaround here is to help bundler by reducing the number of possibilities it needs to consider. In particular, restricting rails usually does the trick, because it brings a lot of complex dependencies into the picture. In the case of https://github.com/rapid7/metasploit_data_models, I see that it depends on `railties ~> 6.0`, `activerecord ~> 6.0`, and `activesupport ~> 6.0`. Changing the Gemfile to also restrict the `rails` dependency does the trick:

```diff
(master) deivid@chaba ~/Code/metasploit_data_models $ git diff
diff --git a/Gemfile b/Gemfile
index b3f4567..e8e8937 100755
--- a/Gemfile
+++ b/Gemfile
@@ -20,7 +20,7 @@ group :development, :test do
   # auto-load factories from spec/factories
   gem 'factory_bot_rails'
 
-  gem 'rails'
+  gem 'rails', '~> 6.0'
   # Used to create fake data
   gem "faker"
 
```

**Comment by duckinator** [comment]
Yeah, that lines up with what I've discovered. The more I look into it, the more I think my mostly-joking comment I made elsewhere yesterday pretty much summed it up:

> package managers are sneaky because one second you're debugging a performance problem and the next you're reading about graph theory because you realized the "performance problem" was actually "doing things correctly" and it turns out "doing things correctly" is pretty slow

I also just confirmed that pinning rails in my testcase will make it finish in 0.6 seconds.

So, personally: I'm not sure if I'd call this "fixed," but if nothing else we have a known workaround for a known problem and it probably should be documented somewhere.

**Comment by duckinator** [comment]
@deivid-rodriguez you're seem at least vaguely familiar with this part of the codebase -- do you know if it's possible to get a ballpark estimate of how complex the set of dependencies/constraints are?

If we can do that, we can perhaps determine a threshold above which we show a warning letting folks know basically "hey, this might take a long time to finish. perhaps you should pin more complex dependencies?" (except phrased better, and bonus points if we can recommend a specific thing to pin).

**Comment by deivid-rodriguez** [comment]
I don't think the complexity of a resolution can be estimated beforehand, at least I'm not sure how. I don't think we should warn or document this either, given how rare this seems at the moment. Resolution is normally quite fast right now.

That said, I don't think we should call this fixed either, and we should aim to fix this properly. It's unacceptable that bundler hangs while resolving, there must be something wrong in this particular case, we need to investigate more.

**Comment by duckinator** [comment]
> there must be something wrong in this particular case

The problem of resolving dependencies can be represented as a variant of the ["boolean satisfiability problem" (aka SAT)](https://en.wikipedia.org/wiki/Boolean_satisfiability_problem).

From the wikipedia page for the boolean satisfiability problem:

> Since the SAT problem is NP-complete, only algorithms with exponential worst-case complexity are known for it. 

This means there will always be poorly-performing cases. But doing things like adding version constraints removes part of the problem set, and can change how close it is to that worst-case scenario. I suspect, in practice, there are possibilities that can be preemptively ruled out for various reasons. I don't know how to find those, unfortunately.

I will also note that Pip has recently redone their resolver, and are running into similar problems: https://github.com/pypa/pip/issues/9187

**Comment by deivid-rodriguez** [comment]
Yeah, sure, but it's not acceptable that most resolutions are very fast, yet we don't even know whether these two cases actually finish at some point or it run into infinite loops. I believe we should be able to do a lot better.

**Comment by duckinator** [comment]
Yeah, agreed &mdash; just wanted to make note that pip is having similar problems, and make sure we're aware of exactly what weeds we're wading into. :slightly_smiling_face:

**Comment by ashkulz** [comment]
This affects us as well, pinned to `2.2.14` to fix the issue. Trying to extract a reproducible test case will be very hard, not sure if it's needed?

**Comment by deivid-rodriguez** [comment]
Did you mean `2.1.4` or `2.2.14`? If this is actually a 2.2.15 specific regression, that would be very useful to know.

**Comment by ashkulz** [comment]
We had earlier pinned to 2.2.1 due to https://github.com/dependabot/dependabot-core/issues/2927 and I ran the following commands (for checking with latest Rails via [bootboot](https://github.com/shopify/bootboot))
```
git reset --hard HEAD
cp Gemfile.lock Gemfile_next.lock
DEPENDENCIES_NEXT=1 time -p bundle _<version>_ update rails actionmailer actionpack actionview activejob activemodel activerecord activesupport railties --conservative --strict --major
```
| version | time taken |
|---------|----------|
| 2.2.1 | 41.85 |
| 2.2.14 | 69.53 |
| 2.2.15 | 72.37 |
| 2.2.16 | 70.14 |
| 2.2.17 | 69.45 |
| 2.2.18 | 54.56 |
| 2.2.19 | 51.00 |
| 2.2.20 | 51.85 |
| 2.2.21 | 51.61 |
| 2.2.22 | (_doesn't complete_) |
| 2.2.23 | (_doesn't complete_) |
| 2.2.24 | (_doesn't complete_) |

I didn't bisect it earlier, looks like 2.2.22 is the culrpit (_at least for our use case_). It ran in our CI for over 6 hours, here I killed it after 5 minutes. Let me know if any more information will help!

**Comment by deivid-rodriguez** [comment]
@ashkulz Definitely helps, indeed. It should be a separate issue, since this ticket is a "resolution from scratch" issue, while your case it's a problem when updating selected gems conservatively. It sounds like this is a regression introduced by https://github.com/rubygems/rubygems/pull/4692. Does it work if you don't use the `--conservative` flag? A reproducible case will be definitely needed to fix this, in any case.

**Comment by ashkulz** [comment]
@deivid-rodriguez like I said, it's at $WORK and can't share the exact file(s). I'll try a stab at making a reproducible test case over the weekend. Using `--strict --major` (without `--conservative`, like you said) still did not complete in 5 minutes with 2.2.22 and 2.2.24.

**Comment by deivid-rodriguez** [comment]
Sure, that reproducible test case would be very helpful. Something else that would be helpful is isolating the exact commit that caused this by running a more finer grained bisection.

**Comment by deivid-rodriguez** [comment]
I moved this discussion to #4776 to keep both issues separate.

**Comment by kbruccoleri** [comment]
I have a similar issue in the latest Bundler 2.2.33. Was attempting to upgrade Rails from 5.2.6 to 6.0.0 and it run continuously for 16 hours before I killed it.

The app is a large mono-repo with 409 gems for reference.

**Comment by deivid-rodriguez** [comment]
Yes, this particular issue still exists in the latest bundler, that's why it's still open. Feel free to post a reproducible example here 👍.

**Comment by justin808** [comment]
Here's a case where > 2.2.0 hangs:
https://github.com/shakacode/react_on_rails/commit/3cdb0875cf4f432d3123d739d1d438fa36d4dc5a

**Comment by deivid-rodriguez** [comment]
#5698 should fix this, both the original case and the simplified case provided by @duckinator. I'm not sure about @justin808's case because I used the `Gemfile` at the referenced commit and it bundled just fine with Bundler 2.3.17.

**Comment by glennposadas** [comment]
2022, I experienced this issue. I remember my Bundler version was `2.3.7`. My bundle install for my very first rails app never progressed even after leaving my mac of 5 hours. Until I decided to do `gem install bundler` to update my Bundler.
Finally, run `bundle install --verbose` again. Voila!

**Comment by jhottenstein** [comment]
We are running into this issue for one of our private gems.  I can probably extract a public repo that can reproduce this if it would help (or test potential fixes).

Using a --conservative flag definitely helped, as did removing the `ruby` platform in the Gemfile.lock

**Comment by deivid-rodriguez** [comment]
Yes, please, we're actively working on fixing slow resolving issues right now, so a public repo that reproduces slow resolutions would be 💯
